File: ios/Packages/SolennixFeatures/Sources/SolennixFeatures/Clients/ViewModels/QuickQuoteViewModel.swift
import Foundation
import UIKit
import SolennixCore
import SolennixNetwork

private struct QuickQuotePDFClientPayload: Encodable {
    let name: String
    let phone: String
    let email: String?
}

private struct QuickQuotePDFProductPayload: Encodable {
    let productId: String?
    let name: String
    let quantity: Double
    let unitPrice: Double
    let discount: Double
}

private struct QuickQuotePDFExtraPayload: Encodable {
    let description: String
    let cost: Double
    let price: Double
    let excludeUtility: Bool
}

private struct QuickQuotePDFRequestPayload: Encodable {
    let client: QuickQuotePDFClientPayload?
    let products: [QuickQuotePDFProductPayload]
    let extras: [QuickQuotePDFExtraPayload]
    let numPeople: Int
    let discount: Double
    let discountType: String
    let requiresInvoice: Bool
    let taxRate: Double
}

@Observable
public final class QuickQuoteViewModel {
    let apiClient: APIClient

    // Form Data
    public var numPeople: Int = 100
    public var clientName: String = ""
    public var clientPhone: String = ""
    public var clientEmail: String = ""
    public var showClientInfo: Bool = false

    // Items
    public var selectedProducts: [EventProduct] = []
    public var extras: [EventExtra] = []

    // Financials
    public var discountType: DiscountType = .percent
    public var discountValue: Double = 0
    public var requiresInvoice: Bool = false
    public var taxRate: Double = 16

    // Source data
    public var availableProducts: [Product] = []
    public var productUnitCosts: [String: Double] = [:]

    public var isLoading: Bool = true

    // PDF
    public var pdfData: Data?
    public var showShareSheet: Bool = false

    public init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    @MainActor
    public func loadData() async {
        isLoading = true
        do {
            let products: [Product] = try await apiClient.getAll(Endpoint.products)
            availableProducts = products.filter { $0.isActive }
        } catch {
            print("Error loading products for quick quote: \(error)")
        }
        isLoading = false
    }

    // Handlers
    public func addProduct() {
        if let first = availableProducts.first {
            selectedProducts.append(EventProduct(
                id: UUID().uuidString,
                eventId: "quick-quote",
                productId: first.id,
                quantity: 1,
                unitPrice: first.basePrice,
                discount: 0,
                createdAt: ""
            ))
            Task {
                await fetchIngredientCost(for: first.id)
            }
        }
    }

    public func removeProduct(at offsets: IndexSet) {
        selectedProducts.remove(atOffsets: offsets)
    }

    public func addExtra() {
        extras.append(EventExtra(
            id: UUID().uuidString,
            eventId: "quick-quote",
            description: "",
            cost: 0,
            price: 0,
            excludeUtility: false,
            createdAt: ""
        ))
    }

    public func removeExtra(at offsets: IndexSet) {
        extras.remove(atOffsets: offsets)
    }

    @MainActor
    private func fetchIngredientCost(for productId: String) async {
        if productUnitCosts[productId] != nil { return }
        do {
            let ingredients: [ProductIngredient] = try await apiClient.get(Endpoint.productIngredients(productId))
            let cost = ingredients
                .filter { $0.type == .ingredient }
                .reduce(0.0) { sum, ing in
                    sum + (ing.quantityRequired * (ing.unitCost ?? 0.0))
                }
            productUnitCosts[productId] = cost
        } catch {
            print("Error fetching ingredient costs: \(error)")
        }
    }

    public func updateCostsForCurrentProducts() {
        for product in selectedProducts {
            Task {
                await fetchIngredientCost(for: product.productId)
            }
        }
    }

    // Financial calculations
    public var financials: FinancialSummary {
        let productsSubtotal = selectedProducts.reduce(0.0) { sum, item in
            sum + ((item.unitPrice - item.discount) * Double(item.quantity))
        }

        let normalExtrasTotal = extras.filter { !$0.excludeUtility }.reduce(0.0) { $0 + $1.price }
        let passThroughExtrasTotal = extras.filter { $0.excludeUtility }.reduce(0.0) { $0 + $1.price }
        let extrasTotal = extras.reduce(0.0) { $0 + $1.price }

        let discountableBase = productsSubtotal + normalExtrasTotal

        let discountAmount: Double
        if discountType == .percent {
            discountAmount = discountableBase * (discountValue / 100)
        } else {
            discountAmount = min(discountValue, discountableBase)
        }

        let discountedBase = discountableBase - discountAmount
        let baseTotal = discountedBase + passThroughExtrasTotal

        let taxAmount = requiresInvoice ? baseTotal * (taxRate / 100) : 0
        let total = baseTotal + taxAmount

        // Costs
        let productCost = selectedProducts.reduce(0.0) { sum, p in
            sum + ((productUnitCosts[p.productId] ?? 0) * Double(p.quantity))
        }
        let extrasCost = extras.reduce(0.0) { $0 + $1.cost }
        let totalCost = productCost + extrasCost

        let revenue = total - (requiresInvoice ? taxAmount : 0)
        let profit = revenue - totalCost

        let passThroughRevenue = passThroughExtrasTotal
        let adjustedRevenue = revenue - passThroughRevenue
        let margin = adjustedRevenue > 0 ? (profit / adjustedRevenue) * 100 : 0

        return FinancialSummary(
            productsSubtotal: productsSubtotal,
            extrasTotal: extrasTotal,
            discountAmount: discountAmount,
            taxAmount: taxAmount,
            total: total,
            productCost: productCost,
            extrasCost: extrasCost,
            totalCost: totalCost,
            profit: profit,
            margin: margin
        )
    }

    // MARK: - PDF Generation

    /// Downloads a backend-generated quick quote PDF using current form data.
    @MainActor
    public func generatePDF() async {
        let validProducts = selectedProducts.filter { !$0.productId.isEmpty && $0.quantity > 0 }
        let validExtras = extras.filter { !$0.description.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }

        guard !validProducts.isEmpty || !validExtras.isEmpty else {
            return
        }

        let payload = QuickQuotePDFRequestPayload(
            client: clientName.isEmpty ? nil : QuickQuotePDFClientPayload(
                name: clientName,
                phone: clientPhone,
                email: clientEmail.isEmpty ? nil : clientEmail
            ),
            products: validProducts.map { product in
                QuickQuotePDFProductPayload(
                    productId: product.productId,
                    name: availableProducts.first(where: { $0.id == product.productId })?.name ?? QuickQuoteStrings.product,
                    quantity: Double(product.quantity),
                    unitPrice: product.unitPrice,
                    discount: product.discount
                )
            },
            extras: validExtras.map { extra in
                QuickQuotePDFExtraPayload(
                    description: extra.description,
                    cost: extra.cost,
                    price: extra.price,
                    excludeUtility: extra.excludeUtility
                )
            },
            numPeople: numPeople,
            discount: discountValue,
            discountType: discountType == .percent ? "percent" : "fixed",
            requiresInvoice: requiresInvoice,
            taxRate: taxRate
        )

        do {
            pdfData = try await apiClient.postData(Endpoint.quickQuotePDF, body: payload)
            showShareSheet = true
        } catch {
            pdfData = nil
            showShareSheet = false
        }
    }
}

public struct FinancialSummary {
    public let productsSubtotal: Double
    public let extrasTotal: Double
    public let discountAmount: Double
    public let taxAmount: Double
    public let total: Double
    public let productCost: Double
    public let extrasCost: Double
    public let totalCost: Double
    public let profit: Double
    public let margin: Double
}

File: android/feature/clients/src/main/java/com/creapolis/solennix/feature/clients/viewmodel/QuickQuoteViewModel.kt
package com.creapolis.solennix.feature.clients.viewmodel

import android.content.Context
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.creapolis.solennix.core.data.repository.ClientRepository
import com.creapolis.solennix.core.data.repository.ProductRepository
import com.creapolis.solennix.core.model.Client
import com.creapolis.solennix.core.model.DiscountType
import com.creapolis.solennix.core.model.InventoryType
import com.creapolis.solennix.core.model.Product
import com.creapolis.solennix.core.model.ProductIngredient
import com.creapolis.solennix.core.network.ApiService
import com.creapolis.solennix.core.network.Endpoints
import com.creapolis.solennix.core.network.get
import com.creapolis.solennix.feature.clients.R
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.io.File
import java.util.UUID
import javax.inject.Inject

data class QuoteItem(
    val id: String = UUID.randomUUID().toString(),
    val productId: String = "",
    val productName: String = "",
    val quantity: Int = 1,
    val unitPrice: Double = 0.0
) {
    val subtotal: Double get() = quantity * unitPrice
}

data class QuoteExtra(
    val id: String = UUID.randomUUID().toString(),
    val description: String = "",
    val price: Double = 0.0,
    val cost: Double = 0.0,
    val excludeUtility: Boolean = false
)

data class QuickQuoteUiState(
    val products: List<Product> = emptyList(),
    val client: Client? = null,
    val selectedItems: List<QuoteItem> = listOf(QuoteItem()),
    val extras: List<QuoteExtra> = emptyList(),
    val numPeople: String = "",
    val requiresInvoice: Boolean = false,
    val taxRate: String = "16",
    val discount: String = "",
    val discountType: DiscountType = DiscountType.PERCENT,
    val productUnitCosts: Map<String, Double> = emptyMap(),
    val isLoading: Boolean = false,
    val isGeneratingPdf: Boolean = false,
    val errorMessage: String? = null,
    val generatedPdfFile: File? = null,
    val clientName: String = "",
    val clientPhone: String = "",
    val clientEmail: String = "",
    val showClientInfo: Boolean = false
) {
    val subtotalProducts: Double
        get() = selectedItems.sumOf { it.subtotal }

    val normalExtrasTotal: Double
        get() = extras.filter { !it.excludeUtility }.sumOf { it.price }

    val passThroughExtrasTotal: Double
        get() = extras.filter { it.excludeUtility }.sumOf { it.price }

    val extrasTotal: Double
        get() = extras.sumOf { it.price }

    val subtotal: Double
        get() = subtotalProducts + extrasTotal

    val discountableBase: Double
        get() = subtotalProducts + normalExtrasTotal

    val discountAmount: Double
        get() {
            val discountValue = discount.toDoubleOrNull() ?: 0.0
            return when (discountType) {
                DiscountType.PERCENT -> discountableBase * discountValue / 100.0
                DiscountType.FIXED -> discountValue.coerceAtMost(discountableBase)
            }
        }

    val taxAmount: Double
        get() {
            if (!requiresInvoice) return 0.0
            val rate = taxRate.toDoubleOrNull() ?: 0.0
            val discountedBase = discountableBase - discountAmount
            return (discountedBase + passThroughExtrasTotal) * rate / 100.0
        }

    val total: Double
        get() {
            val discountedBase = discountableBase - discountAmount
            return discountedBase + passThroughExtrasTotal + taxAmount
        }

    // Profitability
    val costProducts: Double
        get() = selectedItems.sumOf { (productUnitCosts[it.productId] ?: 0.0) * it.quantity }

    val costExtras: Double
        get() = extras.filter { !it.excludeUtility }.sumOf { it.cost }

    val totalCosts: Double
        get() = costProducts + costExtras

    val revenue: Double
        get() = total - taxAmount

    val netProfit: Double
        get() = revenue - totalCosts

    val profitMargin: Double
        get() {
            val adjustedRevenue = revenue - passThroughExtrasTotal
            return if (adjustedRevenue > 0) (netProfit / adjustedRevenue) * 100 else 0.0
        }

    val hasCosts: Boolean
        get() = totalCosts > 0
}

@Serializable
private data class QuickQuotePDFClientPayload(
    val name: String,
    val phone: String,
    val email: String? = null
)

@Serializable
private data class QuickQuotePDFProductPayload(
    @SerialName("product_id") val productId: String? = null,
    val name: String,
    val quantity: Double,
    @SerialName("unit_price") val unitPrice: Double,
    val discount: Double
)

@Serializable
private data class QuickQuotePDFExtraPayload(
    val description: String,
    val cost: Double,
    val price: Double,
    @SerialName("exclude_utility") val excludeUtility: Boolean
)

@Serializable
private data class QuickQuotePDFRequestPayload(
    val client: QuickQuotePDFClientPayload? = null,
    val products: List<QuickQuotePDFProductPayload>,
    val extras: List<QuickQuotePDFExtraPayload>,
    @SerialName("num_people") val numPeople: Int,
    val discount: Double,
    @SerialName("discount_type") val discountType: String,
    @SerialName("requires_invoice") val requiresInvoice: Boolean,
    @SerialName("tax_rate") val taxRate: Double
)

@HiltViewModel
class QuickQuoteViewModel @Inject constructor(
    private val productRepository: ProductRepository,
    private val clientRepository: ClientRepository,
    private val apiService: ApiService,
    @ApplicationContext private val appContext: Context,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val clientId: String? = savedStateHandle["clientId"]

    private val _uiState = MutableStateFlow(QuickQuoteUiState(isLoading = true))
    val uiState: StateFlow<QuickQuoteUiState> = _uiState.asStateFlow()

    init {
        loadData()
    }

    private fun loadData() {
        viewModelScope.launch {
            try {
                // Sync products from network first
                try {
                    productRepository.syncProducts()
                } catch (_: Exception) {
                    // Use cached data if sync fails
                }

                val products = productRepository.getProducts().first()
                val client = clientId?.let { clientRepository.getClient(it) }

                _uiState.update {
                    it.copy(
                        products = products.filter { p -> p.isActive },
                        client = client,
                        isLoading = false
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        errorMessage = e.message ?: appContext.getString(R.string.clients_quick_quote_load_error),
                        isLoading = false
                    )
                }
            }
        }
    }

    fun updateNumPeople(value: String) {
        _uiState.update { it.copy(numPeople = value) }
    }

    fun addItem() {
        _uiState.update { state ->
            state.copy(selectedItems = state.selectedItems + QuoteItem())
        }
    }

    fun removeItem(itemId: String) {
        _uiState.update { state ->
            val items = state.selectedItems.filter { it.id != itemId }
            state.copy(selectedItems = items.ifEmpty { listOf(QuoteItem()) })
        }
    }

    fun updateItemProduct(itemId: String, product: Product) {
        _uiState.update { state ->
            state.copy(
                selectedItems = state.selectedItems.map { item ->
                    if (item.id == itemId) {
                        item.copy(
                            productId = product.id,
                            productName = product.name,
                            unitPrice = product.basePrice
                        )
                    } else item
                }
            )
        }
        fetchProductCosts(product.id)
    }

    fun updateItemQuantity(itemId: String, quantity: String) {
        val qty = quantity.toIntOrNull() ?: return
        if (qty < 1) return
        _uiState.update { state ->
            state.copy(
                selectedItems = state.selectedItems.map { item ->
                    if (item.id == itemId) item.copy(quantity = qty) else item
                }
            )
        }
    }

    fun updateItemPrice(itemId: String, price: String) {
        val p = price.toDoubleOrNull() ?: return
        if (p < 0) return
        _uiState.update { state ->
            state.copy(
                selectedItems = state.selectedItems.map { item ->
                    if (item.id == itemId) item.copy(unitPrice = p) else item
                }
            )
        }
    }

    fun addExtra() {
        _uiState.update { state ->
            state.copy(extras = state.extras + QuoteExtra())
        }
    }

    fun removeExtra(extraId: String) {
        _uiState.update { state ->
            state.copy(extras = state.extras.filter { it.id != extraId })
        }
    }

    fun updateExtraDescription(extraId: String, description: String) {
        _uiState.update { state ->
            state.copy(
                extras = state.extras.map { extra ->
                    if (extra.id == extraId) extra.copy(description = description) else extra
                }
            )
        }
    }

    fun updateExtraPrice(extraId: String, price: String) {
        val p = price.toDoubleOrNull() ?: return
        _uiState.update { state ->
            state.copy(
                extras = state.extras.map { extra ->
                    if (extra.id == extraId) extra.copy(price = p) else extra
                }
            )
        }
    }

    fun updateExtraCost(extraId: String, cost: String) {
        val c = cost.toDoubleOrNull() ?: return
        _uiState.update { state ->
            state.copy(
                extras = state.extras.map { extra ->
                    if (extra.id == extraId) {
                        if (extra.excludeUtility) {
                            extra.copy(cost = c, price = c)
                        } else {
                            extra.copy(cost = c)
                        }
                    } else extra
                }
            )
        }
    }

    fun updateExtraExcludeUtility(extraId: String, excludeUtility: Boolean) {
        _uiState.update { state ->
            state.copy(
                extras = state.extras.map { extra ->
                    if (extra.id == extraId) {
                        if (excludeUtility) {
                            extra.copy(excludeUtility = true, price = extra.cost)
                        } else {
                            extra.copy(excludeUtility = false)
                        }
                    } else extra
                }
            )
        }
    }

    fun updateRequiresInvoice(value: Boolean) {
        _uiState.update { it.copy(requiresInvoice = value) }
    }

    fun updateTaxRate(value: String) {
        _uiState.update { it.copy(taxRate = value) }
    }

    fun updateDiscount(value: String) {
        _uiState.update { it.copy(discount = value) }
    }

    fun updateDiscountType(type: DiscountType) {
        _uiState.update { it.copy(discountType = type) }
    }

    fun updateClientName(value: String) {
        _uiState.update { it.copy(clientName = value) }
    }

    fun updateClientPhone(value: String) {
        _uiState.update { it.copy(clientPhone = value) }
    }

    fun updateClientEmail(value: String) {
        _uiState.update { it.copy(clientEmail = value) }
    }

    fun toggleClientInfo() {
        _uiState.update { it.copy(showClientInfo = !it.showClientInfo) }
    }

    private fun fetchProductCosts(productId: String) {
        if (_uiState.value.productUnitCosts.containsKey(productId)) return
        viewModelScope.launch {
            try {
                val ingredients: List<ProductIngredient> = apiService.get(Endpoints.productIngredients(productId))
                val cost = ingredients
                    .filter { it.type == InventoryType.INGREDIENT || it.type == InventoryType.SUPPLY }
                    .sumOf { (it.unitCost ?: 0.0) * it.quantityRequired }
                _uiState.update { state ->
                    state.copy(productUnitCosts = state.productUnitCosts + (productId to cost))
                }
            } catch (_: Exception) {
                // Silently fail — costs are optional for profitability display
            }
        }
    }

    private fun fetchAllProductCosts() {
        val state = _uiState.value
        state.selectedItems
            .map { it.productId }
            .filter { it.isNotEmpty() }
            .distinct()
            .forEach { fetchProductCosts(it) }
    }

    fun clearPdfFile() {
        _uiState.update { it.copy(generatedPdfFile = null) }
    }

    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    fun generatePdf(context: Context) {
        val state = _uiState.value
        val validItems = state.selectedItems.filter { it.productId.isNotEmpty() && it.quantity > 0 }
        val validExtras = state.extras.filter { it.description.isNotBlank() }

        if (validItems.isEmpty() && validExtras.isEmpty()) {
            _uiState.update { it.copy(errorMessage = appContext.getString(R.string.clients_quick_quote_add_product_error)) }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isGeneratingPdf = true) }
            try {
                val payload = QuickQuotePDFRequestPayload(
                    client = state.client?.let {
                        QuickQuotePDFClientPayload(
                            name = it.name,
                            phone = it.phone,
                            email = it.email
                        )
                    } ?: state.clientName.takeIf { it.isNotBlank() }?.let {
                        QuickQuotePDFClientPayload(
                            name = it,
                            phone = state.clientPhone,
                            email = state.clientEmail.takeIf { email -> email.isNotBlank() }
                        )
                    },
                    products = validItems.map { item ->
                        QuickQuotePDFProductPayload(
                            productId = item.productId,
                            name = item.productName.ifBlank { appContext.getString(R.string.common_product) },
                            quantity = item.quantity.toDouble(),
                            unitPrice = item.unitPrice,
                            discount = 0.0
                        )
                    },
                    extras = validExtras.map { extra ->
                        QuickQuotePDFExtraPayload(
                            description = extra.description,
                            cost = extra.cost,
                            price = extra.price,
                            excludeUtility = extra.excludeUtility
                        )
                    },
                    numPeople = state.numPeople.toIntOrNull() ?: 0,
                    discount = state.discount.toDoubleOrNull() ?: 0.0,
                    discountType = when (state.discountType) {
                        DiscountType.PERCENT -> "percent"
                        DiscountType.FIXED -> "fixed"
                    },
                    requiresInvoice = state.requiresInvoice,
                    taxRate = state.taxRate.toDoubleOrNull() ?: 0.0
                )

                val bytes = apiService.postBytes(Endpoints.QUICK_QUOTES_PDF, payload)
                val file = File(context.cacheDir, "cotizacion_rapida.pdf").apply {
                    writeBytes(bytes)
                }
                _uiState.update { it.copy(generatedPdfFile = file, isGeneratingPdf = false) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        errorMessage = appContext.getString(
                            R.string.clients_quick_quote_pdf_error,
                            e.message ?: e.javaClass.simpleName
                        ),
                        isGeneratingPdf = false
                    )
                }
            }
        }
    }
}
